public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -1,47 +1,56 @@
 import _ from 'lodash';
 import TableModel from 'app/core/table_model';
-import { TimeSeries, FieldType, Labels, formatLabels, QueryResultMeta } from '@grafana/data';
+import { TimeSeries, FieldType, Labels, formatLabels, QueryResultMeta, toDataFrame } from '@grafana/data';
 import { TemplateSrv } from 'app/features/templating/template_srv';
+import { PromDataSuccessResponse, PromMatrixData, TransformOptions } from './types';
+import { FetchResponse } from '@grafana/runtime';
 
 export class ResultTransformer {
   constructor(private templateSrv: TemplateSrv) {}
 
-  transform(response: any, options: any): Array<TableModel | TimeSeries> {
-    const prometheusResult = response.data.data.result;
+  transform(response: FetchResponse<PromDataSuccessResponse>, options: TransformOptions) {
+    const prometheusResult = response.data.data;
+
+    if (!prometheusResult.result) {
+      return [];
+    }
 
     if (options.format === 'table') {
-      return [
-        this.transformMetricDataToTable(
-          prometheusResult,
-          options.responseListLength,
-          options.refId,
-          options.meta,
-          options.valueWithRefId
-        ),
-      ];
-    } else if (prometheusResult && options.format === 'heatmap') {
+      const tableData = this.transformMetricDataToTable(
+        prometheusResult.result,
+        options.responseListLength,
+        options.refId,
+        options.meta,
+        options.valueWithRefId
+      );
+      return [toDataFrame(tableData)];
+    }
+
+    if (options.format === 'heatmap') {
       let seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
+      for (const metricData of (prometheusResult as PromMatrixData).result) {
         seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
       }
       seriesList.sort(sortSeriesByLabel);
       seriesList = this.transformToHistogramOverTime(seriesList);
-      return seriesList;
-    } else if (prometheusResult) {
-      const seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        if (response.data.data.resultType === 'matrix') {
-          seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-        } else if (response.data.data.resultType === 'vector') {
-          seriesList.push(this.transformInstantMetricData(metricData, options));
-        }
+      return seriesList.map(timeSeries => toDataFrame(timeSeries));
+    }
+
+    const seriesList: TimeSeries[] = [];
+
+    if (prometheusResult.resultType === 'matrix') {
+      for (const metricData of prometheusResult.result) {
```

Comment:
I assume if response has 1 `resultType` they cannot be mixed, like having both `vector` and `matrix` in the same response right?

Reply:
no I don't think so.